yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryService.scala

```diff
@@ -0,0 +1,1048 @@
+        val head = _entityQueue.poll(timeLimit - now(), TimeUnit.MILLISECONDS)
+        if (head != null) {
+          val (_, ex) = postOneEntity(head)
+          if (ex.isDefined) {
+            // failure, push back to try again
+            _entityQueue.addFirst(head)
+            if (retryInterval > 0) {
+              Thread.sleep(Math.min(retryInterval, timeLimit - now()))
+            }
+          }
+        }
+      } catch {
+        case ex: InterruptedException =>
+          // ignoring, as this is shutdown time anyway
+          logDebug("Ignoring Interrupt", ex)
+        case ex: Exception =>
+          throw ex
+      }
+    }
+  }
+
+  /**
+   * If the event reaches the batch size or flush is true, push events to ATS.
+   *
+   * @param event event. If null no event is queued, but the post-queue flush logic still applies
+   */
+  private def handleEvent(event: SparkListenerEvent): Unit = {
+    var push = false
+    var isLifecycleEvent = false;
+    val timestamp = now()
+    if (_eventsProcessed.incrementAndGet() % 1000 == 0) {
+      logDebug(s"${_eventsProcessed} events are processed")
+    }
+    event match {
+      case start: SparkListenerApplicationStart =>
+        // we already have all information,
+        // flush it for old one to switch to new one
+        logDebug(s"Handling application start event: $event")
+        if (!appStartEventProcessed.getAndSet(true)) {
+          applicationStartEvent = Some(start)
+          applicationName = start.appName
+          if (applicationName == null || applicationName.isEmpty) {
+            logWarning("Application does not have a name")
+            applicationName = applicationId.toString
+          }
+          userName = start.sparkUser
+          startTime = start.time
+          if (startTime == 0) {
+            startTime = timestamp
+          }
+          setContextAppAndAttemptInfo(start.appId, start.appAttemptId)
+          logInfo(s"Application started: $event")
+          isLifecycleEvent = true
+          push = true
+        } else {
+          logWarning(s"More than one application start event received -ignoring: $start")
+        }
+
+      case end: SparkListenerApplicationEnd =>
+        if (!appStartEventProcessed.get()) {
+          logError(s"Received application end event without application start $event")
+        } else if (!appEndEventProcessed.getAndSet(true)) {
+          // we already have all information,
+          // flush it for old one to switch to new one
+          logInfo(s"Application end event: $event")
+          applicationEndEvent = Some(end);
+          // flush old entity
+          endTime = if (end.time > 0) end.time else timestamp
+          push = true
+          isLifecycleEvent = true
+        } else {
+          logInfo(s"Discarding duplicate application end event $end")
+        }
+
+      case _ =>
+    }
+
+    val tlEvent = toTimelineEvent(event, timestamp)
+    val eventCount = if (canAddEvent(isLifecycleEvent)) {
```

Comment:
This was a tricky one to code. I decided just to block on the number of queued postings, rather than the actual number of queued items. It's easier to measure (you don't need to count the length of postings), and it's a sign of post problems. 

I could switch to doing a count of the number of events in each posting, for a total number: but what would it gain? More succinctly: Is it better to discard excess events because "we have reached our event limit" or because "we have too many pending post operations"?